lib/dynflow/director/execution_plan_manager.rb

```diff
@@ -0,0 +1,109 @@
+module Dynflow
+  class Director
+    class ExecutionPlanManager
+      include Algebrick::TypeCheck
+      include Algebrick::Matching
+
+      attr_reader :execution_plan, :future
+
+      def initialize(world, execution_plan, future)
+        @world                 = Type! world, World
+        @execution_plan        = Type! execution_plan, ExecutionPlan
+        @future                = Type! future, Concurrent::Edge::Future
+        @running_steps_manager = RunningStepsManager.new(world)
+
+        unless [:planned, :paused].include? execution_plan.state
+          raise "execution_plan is not in pending or paused state, it's #{execution_plan.state}"
+        end
+        execution_plan.execution_history.add('start execution', @world.id)
+        execution_plan.update_state(:running)
+      end
+
+      def start
+        raise "The future was already set" if @future.completed?
+        start_run or start_finalize or finish
+      end
+
+      def prepare_next_step(step)
+        StepWorkItem.new(execution_plan.id, step).tap do |work|
+          @running_steps_manager.add(step, work)
+        end
+      end
+
+      # @return [Array<WorkItem>] of Work items to continue with
+      def what_is_next(work)
+        Type! work, WorkItem
+
+        compute_next_from_step =-> step do
+          raise unless @run_manager
+          raise if @run_manager.done?
+
+          next_steps = @run_manager.what_is_next(step)
+          if @run_manager.done?
+            start_finalize or finish
+          else
+            next_steps.map { |s| prepare_next_step(s) }
+          end
+        end
+
+        case work
+        when StepWorkItem
+          step = work.step
+          execution_plan.steps[step.id] = step
+          suspended, work = @running_steps_manager.done(work.step)
+          unless suspended
+            work = compute_next_from_step.call step
+          end
+          work
+        when FinalizeWorkItem
+          raise unless @finalize_manager
+          finish
+        end
+      end
+
+      def event(event)
+        Type! event, Event
+        raise unless event.execution_plan_id == @execution_plan.id
```

Comment:
This (and three raises above) should probably have some kind of message, even if they function just as an assertion.